File: LabTab-Back/src/main/java/cl/labtab/api/models/Order.java
package cl.labtab.api.models;

import cl.labtab.api.common.enums.OrderStatusEnum;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "`order`")
public class Order extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "branch_id", nullable = false)
    private UUID branchId;

    @Column(name = "dine_session_id", nullable = false)
    private UUID dineSessionId;

    @Column(name = "person_id")
    private UUID personId;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 50)
    private OrderStatusEnum status = OrderStatusEnum.PLACED;

    @Column(name = "subtotal", nullable = false, precision = 12, scale = 2)
    private BigDecimal subtotal = BigDecimal.ZERO;

    @Column(name = "total", nullable = false, precision = 12, scale = 2)
    private BigDecimal total = BigDecimal.ZERO;

    @Column(name = "item_count", nullable = false)
    private int itemCount;

    @Column(name = "notes")
    private String notes;

    @Column(name = "channel", length = 20)
    private String channel = "staff";

    public UUID getId() {
        return id;
    }

    public UUID getBranchId() {
        return branchId;
    }

    public void setBranchId(UUID branchId) {
        this.branchId = branchId;
    }

    public UUID getDineSessionId() {
        return dineSessionId;
    }

    public void setDineSessionId(UUID dineSessionId) {
        this.dineSessionId = dineSessionId;
    }

    public UUID getPersonId() {
        return personId;
    }

    public void setPersonId(UUID personId) {
        this.personId = personId;
    }

    public OrderStatusEnum getStatus() {
        return status;
    }

    public void setStatus(OrderStatusEnum status) {
        this.status = status;
    }

    public BigDecimal getSubtotal() {
        return subtotal;
    }

    public void setSubtotal(BigDecimal subtotal) {
        this.subtotal = subtotal;
    }

    public BigDecimal getTotal() {
        return total;
    }

    public void setTotal(BigDecimal total) {
        this.total = total;
    }

    public int getItemCount() {
        return itemCount;
    }

    public void setItemCount(int itemCount) {
        this.itemCount = itemCount;
    }

    public String getNotes() {
        return notes;
    }

    public void setNotes(String notes) {
        this.notes = notes;
    }

    public String getChannel() {
        return channel;
    }

    public void setChannel(String channel) {
        this.channel = channel;
    }

    public void calculateTotals(List<OrderLine> lines) {
        this.subtotal = lines.stream().map(OrderLine::getLineTotal).reduce(BigDecimal.ZERO, BigDecimal::add);
        this.itemCount = lines.stream().mapToInt(OrderLine::getQuantity).sum();
        this.total = this.subtotal;
    }
}

File: labtab_app_v1/lib/features/order/domain/entities/order.dart
class Order {
  final String id;
  final String status;
  final int subtotal;
  final int total;
  final int itemCount;
  final List<OrderLine> lines;
  final String? kitchenTicketId;

  const Order({
    required this.id,
    required this.status,
    required this.subtotal,
    required this.total,
    required this.itemCount,
    this.lines = const [],
    this.kitchenTicketId,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json['id'] as String,
      status: json['status'] as String,
      subtotal: (json['subtotal'] as num).toInt(),
      total: (json['total'] as num).toInt(),
      itemCount: json['itemCount'] as int? ?? 0,
      lines: (json['lines'] as List<dynamic>?)
              ?.map((e) => OrderLine.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      kitchenTicketId: json['kitchenTicketId'] as String?,
    );
  }
}

class OrderLine {
  final String id;
  final String? name;
  final int quantity;
  final String status;
  final String? dishName;
  final int? unitPrice;

  const OrderLine({
    required this.id,
    this.name,
    this.quantity = 1,
    required this.status,
    this.dishName,
    this.unitPrice,
  });

  factory OrderLine.fromJson(Map<String, dynamic> json) {
    return OrderLine(
      id: json['id'] as String,
      name: json['name'] as String?,
      quantity: json['quantity'] as int? ?? 1,
      status: json['status'] as String,
      dishName: json['dishName'] as String?,
      unitPrice: json['unitPrice'] != null
          ? (json['unitPrice'] as num).toInt()
          : null,
    );
  }
}
